src/main/java/com/connexta/store/service/impl/StoreServiceImpl.java

```diff
@@ -25,42 +28,71 @@
 import javax.validation.constraints.Size;
 import lombok.AllArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
+import org.springframework.util.StringUtils;
 
 @Slf4j
 @AllArgsConstructor
 public class StoreServiceImpl implements StoreService {
 
+  private static final String FILE_NAME_METADATA_KEY = "Filename";
+
   @NotBlank private final String retrieveEndpoint;
-  @NotNull private final StorageAdaptor storageAdaptor;
+  @NotNull private final StorageAdaptor fileStorageAdaptor;
+  @NotNull private final StorageAdaptor irmStorageAdaptor;
   @NotNull private final IndexDatasetClient indexDatasetClient;
 
   @Override
-  public @NotNull URI createDataset(
-      @NotNull @Min(1L) @Max(10737418240L) Long fileSize,
-      @NotBlank String mediaType,
-      @NotBlank String fileName,
-      @NotNull InputStream fileInputStream)
-      throws CreateDatasetException, URISyntaxException {
-    final String key = UUID.randomUUID().toString().replace("-", "");
-    storageAdaptor.store(fileSize, mediaType, fileName, fileInputStream, key);
-    return new URI(retrieveEndpoint + key);
+  @NotNull
+  public URI createDataset(
+      @NotNull @Min(1L) @Max(10737418240L) final Long fileSize,
+      @NotBlank final String mediaType,
+      @NotBlank final String fileName,
+      @NotNull final InputStream fileInputStream)
+      throws StoreException, URISyntaxException {
+    final String datasetId = UUID.randomUUID().toString().replace("-", "");
+    fileStorageAdaptor.store(
+        fileSize, mediaType, fileInputStream, datasetId, Map.of(FILE_NAME_METADATA_KEY, fileName));
+    return new URI(retrieveEndpoint + datasetId);
   }
 
   @Override
-  public @NotNull RetrieveResponse retrieveFile(
+  public @NotNull FileRetrieveResponse retrieveFile(
       @Pattern(regexp = "^[0-9a-zA-Z]+$") @Size(min = 32, max = 32) final String datasetId)
       throws RetrieveException {
-    return storageAdaptor.retrieve(datasetId);
+    final StorageAdaptorRetrieveResponse storageAdaptorRetrieveResponse =
+        fileStorageAdaptor.retrieve(datasetId);
+
+    final String fileName =
+        storageAdaptorRetrieveResponse.getMetadata().get(FILE_NAME_METADATA_KEY);
```

Comment:
Encapsulate the knowledge of how the metadata is organized inside the StorageAdaptorRetrieveResponse class. Don't expose the information about the metadata map to other classes. 

```
StorageAdaptorRetrieveResponse

String getFilename() {
   String filename = metadata.get(FILE_NAME_METADATA_KEY);
       if (StringUtils.isEmpty(filename)) {
      throw new RetrieveException(
          String.format(
              "Expected S3 object to have a non-null metadata value for %s",
              FILE_NAME_METADATA_KEY));
     }
    return filename;
}
```